Form — Guidance on where to present a form (#2834)
grommet/hpe-design-system: aries-site/src/pages/templates/forms.mdx

- Announce the max character allowed before the user begins typing.
- Announce when the user has typed 50% of the max allowed.

> ### Exceptions and how to address them

Author: done

  href="#sign-in"
  alt="HPE Design System Form where required and optional can be skipped"
/>

> ## Full-page form guidance

Reviewer: ```suggestion
## Full-page form
```
Can we omit "guidance"? Also, I'm questioning its placement. I feel like it should come before "Exceptions".

Author: honestly the whole placement we went back and forth on this page really need some loving or a god to fix the whole format. There is everything everywhere so I made the main section `full page form` then everything related lives within

Reviewer: Yeah the page does need some help. The top-half seems to have some structure but then it becomes a bit more free for all hah

Author: yeah forsure but also even the top half for example when you look at the dos there are things in there referring to required fields same with the Dont and also another example with required field so would be nice is all of that was in one place

Author: but this is a seperate issue ☝️

- Use full-page forms for create and edit flows, irrespective of the size of the form.
- Use a single-column full-page form instead of multi-column layouts.
- Full-page forms shoud be center-aligned.
- Use full-page when the form is too long.

> - Use Center modal instead of side drawers/layers for secondary inputs.

Reviewer: It's not immediately clear what a "secondary input" is.

Also, a "center modal" is a Layer in grommet speak, so this seems misleading. Maybe rewording to "Use a center modal instead of a side drawer".

Author: yeah secondary input is tricky trying to think how else to word it it would be a button in the form that takes you to another form... 🤔 

Ill change I agree that wording makes sense

Author: changed these

> - Use side layers to display help text, filters, or a form over another form.

Reviewer: This seems to contradict the bullet point above that says not to use a side drawer for a form over a form.

Author: Ill check on this

> ### Full Page Form Example

Reviewer: We could probably scrap this heading and just have the example inline

Author: right good thought ill take out

Author: done

<CardGrid cards={formLayouts} />

### Avoid using multiple column layouts

Single-column layouts create clear paths for users.
Multi-column layouts can inadvertently cause users to skip important fields, input data in the wrong field, or simply prevent progress; all of which can lead users to abandon forms.

<BestPracticeGroup>
  <Example
    bestPractice={{
      type: 'do',
      message:
        'One column layer to create a clear path to completion for the user.',
    }}
  >

>     <ShippingExample />

Reviewer: Seems like the "do" should be the full-page form example and the "don't "should be an altered version of that example that's 2 columns so that it's clear what the difference between the do and don't are

Author: okay I can change that to look like full page form

  href="#sign-in"
  alt="HPE Design System Form where required and optional can be skipped"
/>

> ## Full-page form

Reviewer 2: I'm a little confused how this is being presented. Wasn't the original issue to answer the question of "when should I use a form in a side layer vs. in a full page experience?"

I would expect the title of this section to be "Form presentation" or "Forms in Layers vs. dedicated page" or "Forms in Layer vs. full-page"...

Reviewer 3: You are correct. The initial question was that but up on talking to the designers who had this questions and use cases they were in need of a guidance on how to use full page since we had a forms in layer guidance.  I do like the comparison section name."Forms in Layer vs. full-page" . Do you suggest i should link the forms in layer with in this guidance ?

Reviewer 2: Yes, cross-linking would be great.

> - Use full-page forms for create and edit flows, irrespective of the size of the form.

Reviewer 2: Why?

We should be educating and explaining the rationale and why it makes for a better experience. Otherwise we are just dictating "because I said so." How do we go back and justify/rationalize this guidance and the decisions made?....

> - Use full-page forms for create and edit flows, irrespective of the size of the form.

Reviewer 2: What if I am creating or editing an object with only two fields?

I can't select an object from a list and edit its attributes in a side-layer?

> - Use a single-column full-page form instead of multi-column layouts.

Reviewer 2: Why?

We should be educating and explaining the rationale and why it makes for a better experience. Otherwise we are just dictating "because I said so." How do we go back and justify/rationalize this guidance and the decisions made?....

Reviewer 3: Agreed! i just did the research  and just added the guidance without explaining  the why. Valid point i will update that. thanks

> - Full-page forms shoud be center-aligned.

Reviewer 2: Can we see this? How does it work within Page, PageContent, and in relation to PageHeader?

Author: yeah so I have this example locally but was waiting on mutli-select but I did wrap it in `page` and use `pageHeader` and `pageContent`

> - Use full-page when the form is too long.

Reviewer 2: "too long" seems arbitrary. What is too long?

Reviewer 3: Some research suggested  12 and some 10 so i just put too long lol. 😔 Option "Display no more than 10 form fields with in the same viewport." Or "Minimize the number of form fields so users see within the same viewport." Suggestions ...

> - Use Center modal instead of side drawer to display information without leaving the current context

Reviewer 2: I'm not understanding what this is telling me...

<Example>
  <FullPageFormExample />
</Example>

### Avoid using multiple column layouts

Single-column layouts create clear paths for users.

> Multi-column layouts can inadvertently cause users to skip important fields, input data in the wrong field, or simply prevent progress; all of which can lead users to abandon forms.

Reviewer 2: These do a better job answering the "why."

> Single-column layouts create clear paths for users.
> Multi-column layouts can inadvertently cause users to skip important fields, input data in the wrong field, or simply prevent progress; all of which can lead users to abandon forms.

Reviewer 2: ```suggestion
Single-column layouts create a clear, focused path for users.
Multi-column layouts can cause users to skip important fields, input data in the wrong field, or simply impede efficient progress &mdash; leading to frustration or even abandonment.
```

  href="#sign-in"
  alt="HPE Design System Form where required and optional can be skipped"
/>

## Forms in layers vs. full page 

> 

Author: This is changed

> 

Reviewer 2: @bezAdane and @britt6612 - 

I am going to have to come back to review this. I will provide detailed comments and suggestions when I do.

We are getting closer here, but we are still missing the focus on the user and what is the user trying to accomplish. We need to be speaking to the user problems being solved for and then which tool someone should reach for when solving that problem. How can we tie this content back to representative use cases?

If we look at the [original question](https://www.figma.com/file/Lh01flLH73cFNGZiFxEFth/HPE---Office-Hours?node-id=460%3A6264), it is framed in use cases. "In these types of situations, I think we should use A, because of x, y, z reasons/benefits. In this of other type of situation, we should use B because of m, n, o benefits. We need to be presenting our content in this sort of rational flow --> "In this type of situation where the user is trying to accomplish _____, we recommend using this tool/variant/etc., because it provides these benefits and value to the end user."

  href="#sign-in"
  alt="HPE Design System Form where required and optional can be skipped"
/>

## Forms in layers vs. full page 

> - Single column form layout creates a clear focused path to completion for the user.
> - Multi-column layouts can cause users to skip important fields, input data in a wrong field, or simply hinder efficiency which leads to frustration or even abandonment. 

Reviewer 2: ```suggestion
- Single column form layout creates a clear focused path to completion for the user.
- Multi-column layouts can cause users to skip important fields, input data in a wrong field, or simply hinder efficiency which leads to frustration or even abandonment. 
```

These are repeated below. They seem to flow well within the "avoid multi columns"

Author: done my mistake

> - Display no more than 10 form fields with in the same viewport so that users are not intimidated with seeing a long page filled with form fields and sections.

Reviewer 2: Do we have a solution/recommendation we can suggest for use cases where there are a large number of fields? Should we point them to Wizard?

"Consider presenting the [form in a wizard](/templates/wizards) to break the form in to digestible, focussed sections."

Author: This was updated

> - Use a [Center modals](https://design-system.hpe.design/components/layer#center-modal) instead of side drawer to display information without leaving the current contexts. 

Reviewer 2: I am still unsure as to what this is telling me / why this is being recommended. Is there a visual someone could share?

This reads as if I am never supposed to use a side drawer for a form. Under which circumstances should I use a side drawer vs. full page vs. center modal? Perhaps this gets clearer one the "forms in layer vs. full-page" content is written?

Author: content was changed

  href="#sign-in"
  alt="HPE Design System Form where required and optional can be skipped"
/>

## Forms in layers vs. full page 

A full-page form is used to convey information without the possibility of several interpretations. Depending on the type of form we are designing, a well-structured form on how and when to use forms will guide users with minimal dispute. 

**Best Practices**

- When using a full-page form use a single-column form layout to create a clear focused path to complete for the user.
- When using a full-page form avoid using Multi-column layouts not to cause users to skip important fields, input data in the wrong field, or simply hinder efficiency which leads to frustration or even abandonment.
- When using a single form-page use no more than 10 form fields in the same viewport so that users are not intimidated by seeing a long page filled with form fields and sections.

> - Use [center modals](https://design-system.hpe.design/components/layer#center-modal) for full-page forms to display information without leaving the current contexts. 

Reviewer: Wondering if it would be clearer to separate this list into two sections:

Best practices for full-page forms:
- Use a single-column...
- Avoid using multi-column..

Best practices for forms in layers:
- Use center modals...
- Use side drawer...
...

Author: good point I agree would be easier to scan

Reviewer 3: good point, i can update the figma

  href="#sign-in"
  alt="HPE Design System Form where required and optional can be skipped"
/>

## Where to present a form

There are several placement options when presenting a form, but when should a particular 
placement be chosen and why? On a dedicated page? Within a layer? If within a layer, which 
type?

<FormPresentationDecisionTree />

## Forms in layers vs. full page 

A full-page form is used to display information free from interuption.

> **Best practices for full-page forms:**

Author: I would vote to keep all of these bullet points below..

- When using a full-page form use a single-column form layout to create a clear focused path to complete for the user.
- When using a full-page form avoid using Multi-column layouts not to cause users to skip important fields, input data in the wrong field, or simply hinder efficiency which leads to frustration or even abandonment.
- When using a single form-page use no more than 10 form fields in the same viewport so that users are not intimidated by seeing a long page filled with form fields and sections.

> **Best practices for forms in layers:**

Author: In my opinion, these are kinda stated above so I would take them out unless we kinda want them to restate what the diagram says but I don't think it would be bad if we removed

Reviewer: Agree to removing.

Reviewer: To me the "forms in layer vs full page" is confusing because it feels redundant with the decision tree.

Author: agree
